### DISCUSSION TITLE
How to add a sparse loss during training

### DISCUSSION THREAD

**Discussion by rogercmq** [open-mmlab/mmaction2#632] [General] [answered]
I am working on adding a L1-norm loss term during training, aiming to obtain sparse network weights.  Could you give me some advice?

**Comment by congee524** [comment, marked as answer]
Since we calculated loss in the `HEAD`, I think the best way is to modify the optimizer.....

For example, `torch.optim.SGD` directly computes the gradient of L2 regularization (1/2 w^2   --->  w):
```
if weight_decay != 0:
    d_p = d_p.add(p, alpha=weight_decay)
```
Maybe you can
```
d_p = d_p.add(sign(p), alpha=weight_decay_l1)
```

**Comment by dreamerlin** [comment]
It has been answered @rogercmq

**Comment by rogercmq** [comment]
Problem solved. Thx!
